docs/content/docs/deployment/speculative_execution.md

```diff
@@ -0,0 +1,100 @@
+Controls the maximum number of execution attempts of each operator that can execute concurrently, including the original one and speculative ones.
+- [`jobmanager.adaptive-batch-scheduler.speculative.block-slow-node-duration`]({{< ref "docs/deployment/config" >}}#jobmanager-adaptive-batch-scheduler-speculative-block-slow-node): 
+Controls how long an detected slow node should be blocked for.
+
+You can also need to tune below configuration options to control the slow task detector:
+- [`slow-task-detector.check-interval`]({{< ref "docs/deployment/config" >}}#slow-task-detector-check-interval): 
+The interval to check slow tasks.
+- [`slow-task-detector.execution-time.baseline-lower-bound`]({{< ref "docs/deployment/config" >}}#slow-task-detector-execution-time-baseline-lower-bound): 
+The lower bound of slow task detection baseline.
+- [`slow-task-detector.execution-time.baseline-multiplier`]({{< ref "docs/deployment/config" >}}#slow-task-detector-execution-time-baseline-multiplier): 
+The multiplier to calculate the slow tasks detection baseline. Given that the parallelism is N and the ratio is R, define T as the median of the first N*R finished tasks' execution time. The baseline will be T*M, where M is the multiplier of the baseline.
+- [`slow-task-detector.execution-time.baseline-ratio`]({{< ref "docs/deployment/config" >}}#slow-task-detector-execution-time-baseline-ratio): 
+The finished execution ratio threshold to calculate the slow tasks detection baseline. Given that the parallelism is N and the ratio is R, define T as the median of the first N*R finished tasks' execution time. The baseline will be T*M, where M is the multiplier of the baseline.
+
+### Enable Sources for Speculative Execution
+If your job uses a custom {{< gh_link file="/flink-core/src/main/java/org/apache/flink/api/connector/source/Source.java" name="Source" >}}, 
+and the source uses custom {{< gh_link file="/flink-core/src/main/java/org/apache/flink/api/connector/source/SourceEvent.java" name="SourceEvent" >}},
+you need to change the {{< gh_link file="/flink-core/src/main/java/org/apache/flink/api/connector/source/SplitEnumerator.java" name="SplitEnumerator" >}} 
+of that source to implement {{< gh_link file="/flink-core/src/main/java/org/apache/flink/api/connector/source/SupportsHandleExecutionAttemptSourceEvent.java" name="SupportsHandleExecutionAttemptSourceEvent" >}} 
+interface.
+```java
+public interface SupportsHandleExecutionAttemptSourceEvent {
+    void handleSourceEvent(int subtaskId, int attemptNumber, SourceEvent sourceEvent);
+}
+```
+This means the SplitEnumerator should be aware of the attempt which sends the event. Otherwise, exceptions 
+will happen when the job manager receives a source event from the tasks and lead to job failures.
+
+No extra change is required for other sources to work with speculative execution. All the source connectors 
```

Comment:
Maybe describe the `other sources` more clearly, the `other sources` includes  SourceFunction, InputFormat sources and the new sources.